### ISSUE TITLE
Refactor importBlock and ZkWorldStateArchive persistence along finality

### ISSUE THREAD

**Issue by garyschulte** [Consensys/shomei#104]
### Description
Currently when shomei imports blocks, we persist to database storage directly.  There is a specific option `--enable-finalized-block-limit` that prevents the import of blocks that are not finalized.  Currently the finalized block limit/hash is updated via `rollup_forkChoiceUpdated`

In order to be more performant and offer additional functionality, this ticket aims to
1. refactor the finalized block limit feature to be an 'always on' import block limit condition, and the `--enable-finalized-block-limit` cli option should implement a condition similar to today by hash/number.  The default behavior should just return true.  We will refer to this as the "block limit" in this issue

2. import all blocks as they arrive, using the persisted worldstate if the block limit rule returns true, and an in memory worldstate if the block limit returns false.  Additionally, we want to use persistent storage for trace generation so that if we generate the trace against an in-memory worldstate we persist the trace in the db along with the trie log.

3. refactor the ZkWorldStateArchive such that it expects to cache all ZkWorldStates as they arrive.

4. on persisted state rolling, ensure we aggregate all of the in-memory changes into a single commit in order to move to the finalized worldstate.  

The net result should be that we:
* speculatively import shomei blocks
* persist only finalized blocks
* persist traces for all blocks (if configured to trace)
* can quickly roll to any worldstate that is at or ahead of the most recent finalized block.

**Comment by jpnovais** [comment]
It's my understanding that Shomei has 2 operation modes: 
- **A**: for serving `rollup_getZkEVMStateMerkleProofV0(startBlockNumber, endBlockNumber)` with state transition that is proven the prover
- **B**: rollup_getProof(address, ...) (similar of eth_getProof), which is public exposed to the daps/protocols. 

At deployment time we need to choose either A or B, IIRC `--enable-trace-generation=true` runs A,  otherwise is B.

The flag `--enable-finalized-block-limit=true` and `rollup_forkChoiceUpdated` methods are at the moment used only in B mode.

I am not familiar with Shomei internals to fully understand what are the implications of points 1..4, but my understanding is that:

- in both modes A and B, shomei will "optimistically" import blocks as they arrive;
- AFIR atm, rollup_forkChoiceUpdated only allow to set the `finalizedBlockNumber/Hash` and move it forward. To support rollback, will this method now have extra parameters `headBlockNumber/Hash `? How will the rollback be done?
- Is there any operational implications of the mode A, will coordinator now need to call engine_forkchoiceUpdated with finalized block number as well?

**Comment by garyschulte** [comment]
@matkt implemented the current finalized-block-limit feature used for rollup_getProof operation.  What it does is prevent importing anything optimistically by preventing trielogs/blocks PUSHED from besu that are after the finalized block hash/number.

when a new finalized block is set, any subsequent PUSHED block triggers shomei to PULL all of the blocks up until the latest finalized block.  So in this mode, besu is optimistically importing, but shomei is not .

I think this PR is an opportunity to make shomei operate optimistically and faster in the `rollup_getProof`  mode, rather than having to replay all of the blocks in a burst on every finalized block.

I suspect we might be able to use a single shomei node type for both rollup_getProof and rollup_getZkEVMStateMerkleProofV0 operation.  If we structure the worldstate persistence in this way, I don't think we actually need the distinction between the two modes, and would always operate in the "B" mode (unless there is no forkchoice provider).   I will defer to @matkt to check me on this though.

**Comment by garyschulte** [comment]
similarly, if we only persist finalized states, but optimisitcally import all blocks, that should prevent needing to resync shomei if we optimistically import bad blocks.  There would be no rollback, just a new optimistic block that rolls-forward from the last finalized state.

**Comment by jpnovais** [comment]
> I suspect we might be able to use a single shomei node type for both rollup_getProof and rollup_getZkEVMStateMerkleProofV0 operation

If we can achieve that and make it work reliably, it would be great! It would simplify the architecture and maintenance costs.

**Comment by matkt** [comment]
Yes, the idea of the ticket is to have the same behavior for both shomeis, thus eliminating the need for two modes. However, there will be a flag to indicate whether we want to generate zk traces or not (because size of the zktrace is really big), and that will be the only difference. I think by default it will be set to false, and if we want to enable it, we will need to add that.

However, to implement this, the shomei generating the zk traces must receive a rollup_forkChoiceUpdated, which is currently not the case if I remember correctly .

Indeed, the finalized-block-limit prevents shomei from importing a block and its trielog after this limit while waiting for the new limit. The change will therefore be to always import after this limit but only in memory, waiting for the new limit to persist those in memory.

I would say that the proposal we are making will require some refactoring in the sync logic but could address the various needs related to shomei that we have received recently.

**Comment by jpnovais** [comment]
> The change will therefore be to always import after this limit but only in memory, waiting for the new limit to persist those in memory.

What are the memory consumption and performance implications of this? 
We aim to have Finalization on L1 within 36 Hours, however we can be multiple days behind the L2 chain head, especially if have a bug in the system where we have already been 4+ days behind. We must be careful of this scenario.

**Comment by garyschulte** [comment]
> What are the memory consumption and performance implications of this? 
> We aim to have Finalization on L1 within 36 Hours, however we can be multiple days behind the L2 chain head...
> 

We can recreate the in memory states fairly cheaply, so the expectation is that we would only keep a limited number of states in memory.  Just like besu, we only keep 512 states in memory but since we persist trielogs, we can still roll the head to any arbitrary point by applying trielogs on top of the persisted state.  The only difference is that shomei will only be able to roll forward rather than backward like besu. 

Memory should not be a concern and performance should be better for finalization if we:
 * do not create traces at each block
 * Roll the state in batches rather than persisting at each block as we reapply the trielog.
